src/crashreporter/crashuploader.cpp

```diff
@@ -0,0 +1,148 @@
+  m_files = files;
+  nextUpload();
+}
+
+void CrashUploader::nextUpload() {
+  if (!m_files.empty()) {
+    m_currentFile = m_files.first();
+    m_files.pop_front();
+    m_retries = 0;
+    startRequest(m_currentFile);
+  } else {
+    emit uploadsComplete();
+  }
+}
+
+void CrashUploader::startRequest(const QString& file) {
+  logger.debug() << "Starting upload of " << file;
+  QFile dump(file);
+  if (!dump.open(QIODevice::ReadOnly)) {
+    // fail and try the next one
+    logger.error() << "Unable to open file: " << file;
+    nextUpload();
+    return;
+  }
+  m_retries++;
+  if (m_retries > MAX_RETRIES) {
+    nextUpload();
+    return;
+  }
+  QString fileName = file.split("\\").last();
+  QByteArray dumpContent = dump.readAll();
+
+  QHttpMultiPart* multipart = new QHttpMultiPart(this);
+  multipart->setBoundary(QString(BOUNDARY).toLocal8Bit());
+  QHttpPart formPart;
+  QString contentDisp = QString(
+                            "form-data; name=\"upload_file_minidump\"; "
+                            "filename=\"%1\"")
+                            .arg(fileName);
+  formPart.setHeader(QNetworkRequest::ContentDispositionHeader,
+                     QVariant(contentDisp));
+  formPart.setHeader(QNetworkRequest::ContentLengthHeader,
+                     QVariant(dumpContent.size()));
+  formPart.setBody(dumpContent);
+  QHttpPart namePart;
+  namePart.setHeader(QNetworkRequest::ContentDispositionHeader,
+                     "form-data; name=\"ProductName\"");
+  namePart.setBody(QString("MozillaVPN").toLocal8Bit());
+
+  QHttpPart versionPart;
+  versionPart.setHeader(QNetworkRequest::ContentDispositionHeader,
+                        "form-data; name=\"Version\"");
+  versionPart.setBody(APP_VERSION);
+  multipart->append(namePart);
+  multipart->append(formPart);
+  multipart->append(versionPart);
+
+  QString urlStr = SettingsHolder::instance()->stagingServer()
+                       ? Constants::CRASH_STAGING_URL
+                       : Constants::CRASH_PRODUCTION_URL;
+#ifdef MVPN_DEBUG
+  urlStr = Constants::CRASH_STAGING_URL;
+#endif  // MVPN_DEBUG
+
+  logger.debug() << "Uploading to : " << urlStr;
+  QUrl url(Constants::CRASH_PRODUCTION_URL);
+  QNetworkRequest request(url);
+  request.setHeader(QNetworkRequest::UserAgentHeader, "mozillaVPN");
+  // Qt changes the content type automatically to multipart/mixed.  Socorro
+  // doesn't support that, so we have to set it ourselves and create a custom
+  // bounday.
+  request.setHeader(QNetworkRequest::ContentTypeHeader,
+                    QString("multipart/form-data; boundary=%1").arg(BOUNDARY));
+  auto reply = m_network->post(request, multipart);
+
+  connect(reply, &QNetworkReply::sslErrors, [](QList<QSslError> errors) {
+    cout << "SSL Errors: " << endl;
+    for (auto err : errors) {
+      std::cout << err.errorString().toStdString() << std::endl;
+    }
+  });
+  QCoreApplication::processEvents(QEventLoop::ExcludeUserInputEvents);
```

Comment:
Why this? You can use signals and slots to know when the upload is completed and make the code async.